src/v/cluster/topic_table.cc

```diff
@@ -40,26 +40,33 @@ topic_table::transform_topics(Func&& f) const {
 topic_table::topic_metadata::topic_metadata(
   topic_configuration_assignment c, model::revision_id rid) noexcept
   : configuration(std::move(c))
-  , _id_or_topic(rid) {}
+  , _source_topic(std::nullopt)
+  , _revision(rid) {}
 
 topic_table::topic_metadata::topic_metadata(
-  topic_configuration_assignment c, model::topic st) noexcept
+  topic_configuration_assignment c,
+  model::revision_id rid,
+  model::topic st) noexcept
   : configuration(std::move(c))
-  , _id_or_topic(std::move(st)) {}
+  , _source_topic(st)
+  , _revision(rid) {}
 
 bool topic_table::topic_metadata::is_topic_replicable() const {
-    return std::holds_alternative<model::revision_id>(_id_or_topic);
+    return _source_topic.has_value() == false;
 }
+
 model::revision_id topic_table::topic_metadata::get_revision() const {
     vassert(
       is_topic_replicable(), "Query for revision_id on a non-replicable topic");
-    return std::get<model::revision_id>(_id_or_topic);
+    return _revision;
 }
```

Comment:
mm, ok I may be a little tired, but something doesn't seem quite right.

the constructor for the topic_metadata was updated so that when creating a non-replicated topic a revision is stored:

```
      {new_non_rep_topic,
        topic_metadata(std::move(ca), model::revision_id(o()), source.tp)});
```

but then `topic_table::topic_metadata::get_revision()` prevents access to the revision unless it's a replicated topic (via `assert(is_topic_replicable)`. so it's not clear why we'd need to store the revision if we can't access it (and thus why we moved away from std::variant).

Reply:
I'm sure it's something simple I'm missing or maybe it's cleared up in a subsequent commit.

Reply:
As I understand, previously the materialised topics didn't have the revision because there was no controller log command for them. So the code looked exactly like this is still the case. Now we have a command that creates such topics so we can derive the revision id out of the command offset (as it's done with regular topics).
Initially, I removed the assertion because revision exists for both types of topics. I reintroduced the assertion just to keep the old behaviour. When the revision will be needed it will be easy to remove the assertion.

Reply:
Yup Eugene is right here. This is actually a change I should have made once I modified how `non_replicated` topics are created. Furthermore I think its OK to leave out the assertion because (after performing a quick grep) there are currently no callers to `cluster::topic_metadata::get_revision_id()` from coproc anyway. When coproc needs the revision id it actually queries it via log config.